src/zkevm_specs/evm/execution/slt_sgt.py

```diff
@@ -0,0 +1,63 @@
+from typing import Sequence, Tuple
+
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+
+
+def scmp(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    is_sgt, _ = instruction.pair_select(opcode, Opcode.SGT, Opcode.SLT)
+
+    a = instruction.stack_pop()
+    b = instruction.stack_pop()
+    c = instruction.stack_push()
+
+    # decode RLC to bytes for a and b
+    # Note: `rlc_to_bytes` returns the bytes in big endian form. In order to
+    # match the specs and circuit implementation, we transform them to
+    # little endian form (which is the default form returned by `query_word`
+    a8s = instruction.rlc_to_bytes(a, 32)
+    b8s = instruction.rlc_to_bytes(b, 32)
+    c8s = instruction.rlc_to_bytes(c, 32)
+    a8s.reverse()
+    b8s.reverse()
+    c8s.reverse()
+
+    cc = int.from_bytes(c8s, "little")
+    a_hi = int.from_bytes(a8s[16:32], "little")
+    a_lo = int.from_bytes(a8s[0:16], "little")
+    b_hi = int.from_bytes(b8s[16:32], "little")
+    b_lo = int.from_bytes(b8s[0:16], "little")
+
+    a_lt_b = (a_hi < b_hi) or ((a_hi == b_hi) and (a_lo < b_lo))
```

Comment:
I have doubts about this approach of writing the python spec.  Here you're building an expression using `or`, `and`, `<` and `==` operators, which aren't available for building halo2 expressions.

So it seems to me that this spec generates constraints dynamically based on the witness data (a, b), but AFAIK the constraints should be fixed, and all conditional logic should be build with arithmetic expressions.  So I just want to verify if this is the correct way.  

@CPerezz @han0110 what are your thoughts?

Reply:
Hmmm.  I partially agree on that TBH. In the sense that we should mirror the circuit behaviour instead of going over it.

At the same time, we do a lot of byte conversions and similar stuff, which also do not match what happens inside of the circuits 100%.

@han0110 any strong opinions on that? I'm not against it as in the PR implementation in https://github.com/appliedzkp/zkevm-circuits/pull/280 everything is actually explained.

Reply:
After discussing with Han, we are not sure if we should require contributors to write the spec exactly like a circuit because it may be a bit annoying, and actually less clear on what's going on.

In `src/zkevm_specs/evm/instruction.py` -> `Instruction` we have a bunch of methods that act as circuit gadgets.  For example `is_equal` represents the `IsEqualGadget` from the circuits.  This allows showing how gadgets will be composed when writing the spec.

Not all gadgets are found in `Instruction`, but they should be easy to add.  Maybe this line:
```
a_lt_b = (a_hi < b_hi) or ((a_hi == b_hi) and (a_lo < b_lo))
```
Could be rewritten using `instruction.is_lt`, `instruction.is_equal`?

Reply:
For comparison, I added it in #88, perhaps you could just pick and use.

https://github.com/appliedzkp/zkevm-specs/blob/dd061107981ee27ae6e2ab81f78d1b2bb075dc9e/src/zkevm_specs/evm/instruction.py#L229-L232

which takes an extra input `n_bytes ` to specify how many bytes the inputs could be, for underlying range check to work. It intends to reflect how [`ComparisonGadget`](https://github.com/appliedzkp/zkevm-circuits/blob/603409ac6ec1add91345dc6c84355bf0d6d55014/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L608-L617) actually works.

Reply:
@icemelon what do you think?

Reply:
Yes, I agree that using `instruction.compare` is better. 

@roynalnaruto Could you update the PR accordingly?

Reply:
Updating the PR now 👍

Reply:
@icemelon @ed255 Needs approval to re-run the CI build.